File: backend/schemas/fleet.py
"""
schemas/fleet.py
Fleet Management System

Truck and Trailer Pydantic schemas.
"""

from __future__ import annotations
from datetime import datetime
from typing import Optional
from schemas.common import CamelBase, TruckStatus, TrailerStatus, TrailerType, FuelType, VehicleDocumentType


# ─────────────────────────────────────────────────────────────────────────────
# TRUCKS
# ─────────────────────────────────────────────────────────────────────────────

class TruckBase(CamelBase):
    plate_number:           str
    make:                   str
    model:                  str
    year:                   int
    status:                 TruckStatus
    odometer_km:            float
    fuel_type:              FuelType
    vin:                    Optional[str] = None
    color:                  Optional[str] = None
    insurance_expiry_date:  Optional[datetime] = None
    inspection_expiry_date: Optional[datetime] = None
    notes:                  Optional[str] = None


class TruckCreate(TruckBase):
    """POST /fleet/trucks — ADMIN only"""
    pass


class TruckUpdate(CamelBase):
    """PATCH /fleet/trucks/{id} — ADMIN only"""
    plate_number:           Optional[str]         = None
    make:                   Optional[str]         = None
    model:                  Optional[str]         = None
    year:                   Optional[int]         = None
    status:                 Optional[TruckStatus] = None
    odometer_km:            Optional[float]       = None
    fuel_type:              Optional[FuelType]    = None
    vin:                    Optional[str]         = None
    color:                  Optional[str]         = None
    insurance_expiry_date:  Optional[datetime]    = None
    inspection_expiry_date: Optional[datetime]    = None
    notes:                  Optional[str]         = None


class TruckResponse(TruckBase):
    """GET /fleet/trucks  |  GET /fleet/trucks/{id}"""
    id:                 str
    assigned_driver_id: Optional[str] = None
    current_trip_id:    Optional[str] = None
    created_at:         datetime
    updated_at:         datetime


class TruckDocumentResponse(CamelBase):
    """GET /fleet/trucks/{id}/documents"""
    id:          str
    truck_id:    str
    type:        VehicleDocumentType
    file_name:   str
    file_url:    str
    expiry_date: Optional[datetime] = None
    uploaded_at: datetime
    uploaded_by: str


class ServiceRecordResponse(CamelBase):
    """GET /fleet/trucks/{id}/service-history"""
    id:                  str
    truck_id:            str
    service_type:        str
    description:         str
    odometer_at_service: float
    cost:                float
    performed_by:        str
    work_order_id:       Optional[str] = None
    service_date:        datetime
    created_at:          datetime


# ─────────────────────────────────────────────────────────────────────────────
# TRAILERS
# ─────────────────────────────────────────────────────────────────────────────

class TrailerBase(CamelBase):
    plate_number:           str
    make:                   str
    model:                  str
    year:                   int
    status:                 TrailerStatus
    type:                   TrailerType
    capacity_tons:          Optional[float]   = None
    insurance_expiry_date:  Optional[datetime] = None
    inspection_expiry_date: Optional[datetime] = None
    notes:                  Optional[str]     = None


class TrailerCreate(TrailerBase):
    """POST /fleet/trailers — ADMIN only"""
    pass


class TrailerUpdate(CamelBase):
    """PATCH /fleet/trailers/{id} — ADMIN only"""
    plate_number:           Optional[str]           = None
    make:                   Optional[str]           = None
    model:                  Optional[str]           = None
    year:                   Optional[int]           = None
    status:                 Optional[TrailerStatus] = None
    type:                   Optional[TrailerType]   = None
    capacity_tons:          Optional[float]         = None
    insurance_expiry_date:  Optional[datetime]      = None
    inspection_expiry_date: Optional[datetime]      = None
    notes:                  Optional[str]           = None


class TrailerResponse(TrailerBase):
    """GET /fleet/trailers  |  GET /fleet/trailers/{id}"""
    id:               str
    assigned_trip_id: Optional[str] = None
    created_at:       datetime
    updated_at:       datetime


class TrailerDocumentResponse(CamelBase):
    id:          str
    trailer_id:  str
    type:        VehicleDocumentType
    file_name:   str
    file_url:    str
    expiry_date: Optional[datetime] = None
    uploaded_at: datetime
    uploaded_by: str

class FleetSummary(CamelBase):
    total_trucks:      int
    active_trucks:     int
    in_progress_trucks: int
    inactive_trucks:   int
    total_trailers:    int
    active_trailers:   int
    inactive_trailers: int

File: frontend/web/src/types/fleet.ts
/**
 * types/fleet.ts
 * Fleet Management System — Phase 2
 *
 * Changes (Stage 2):
 *   - Truck: added wheelConfig, grossWeightTons, axleLoadTons (optional catalog spec fields)
 *   - Trailer: added axles (optional catalog spec field)
 */

// ─────────────────────────────────────────────────────────────────────────────
// TRUCKS
// ─────────────────────────────────────────────────────────────────────────────

export type TruckStatus = "active" | "inactive" | "in-progress" | "under-maintenance";

export interface Truck {
  id:                    string;
  plateNumber:           string;
  make:                  string;
  model:                 string;
  year:                  number;
  status:                TruckStatus;
  odometerKm:            number;
  fuelType:              "diesel" | "petrol" | "electric" | "hybrid";
  vin?:                  string;
  color?:                string;
  assignedDriverId?:     string;
  currentTripId?:        string;
  insuranceExpiryDate?:  string;   // ISO 8601
  inspectionExpiryDate?: string;
  notes?:                string;
  // Stage 2: catalog spec fields
  wheelConfig?:          string;
  grossWeightTons?:      number;
  axleLoadTons?:         number;
  // Stage 4: vehicle image
  imageUrl?:             string;
  createdAt:             string;
  updatedAt:             string;
}

export interface TruckDocument {
  id:          string;
  truckId:     string;
  type:        "insurance" | "registration" | "inspection" | "other";
  fileName:    string;
  fileUrl:     string;
  expiryDate?: string;
  uploadedAt:  string;
  uploadedBy:  string;
}

export interface ServiceRecord {
  id:                string;
  truckId:           string;
  serviceType:       string;
  description:       string;
  odometerAtService: number;
  cost:              number;
  performedBy:       string;
  workOrderId?:      string;
  serviceDate:       string;
  createdAt:         string;
}

// ─────────────────────────────────────────────────────────────────────────────
// TRAILERS
// ─────────────────────────────────────────────────────────────────────────────

export type TrailerStatus = "active" | "inactive" | "under-maintenance";

export interface Trailer {
  id:                    string;
  plateNumber:           string;
  make:                  string;
  model:                 string;
  year:                  number;
  status:                TrailerStatus;
  type:                  "flatbed" | "refrigerated" | "tanker" | "box" | "other";
  capacityTons?:         number;
  assignedTripId?:       string;
  insuranceExpiryDate?:  string;
  inspectionExpiryDate?: string;
  notes?:                string;
  // Stage 2: catalog spec field
  axles?:                number;
  // Stage 4: vehicle image
  imageUrl?:             string;
  createdAt:             string;
  updatedAt:             string;
}

export interface TrailerDocument {
  id:          string;
  trailerId:   string;
  type:        "insurance" | "registration" | "inspection" | "other";
  fileName:    string;
  fileUrl:     string;
  expiryDate?: string;
  uploadedAt:  string;
  uploadedBy:  string;
}

export interface FleetSummary {
  totalTrucks:      number;
  activeTrucks:     number;
  inProgressTrucks: number;
  inactiveTrucks:   number;
  totalTrailers:    number;
  activeTrailers:   number;
  inactiveTrailers: number;
}